File: src/data_loader.rs
use chrono::prelude::*;
use log::{error, info};
use reqwest::Client;
use serde::Deserialize;
use thiserror::Error;

#[derive(Deserialize)]
struct Point {
    y: f32,
}

#[derive(Deserialize)]
struct DataLine {
    title: String,
    point: Vec<Point>,
}

#[derive(Deserialize)]
struct Data {
    #[serde(rename = "dataLine")]
    data_line: Vec<DataLine>,
}

#[derive(Deserialize)]
struct Response {
    data: Data,
}

#[derive(Error, Debug)]
pub(crate) enum FetchError {
    #[error("Network error: {0}")]
    Network(#[from] reqwest::Error),
    #[error("JSON parsing error: {0}")]
    Json(#[from] serde_json::Error),
    #[error("Price data not found")]
    PriceDataNotFound,
    #[error("Unexpected response status: {0}")]
    UnexpectedStatus(reqwest::StatusCode),
    #[error("Price data does not contain exactly 24 points")]
    InvalidDataSize,
}

pub async fn fetch_data(date: NaiveDate) -> Result<Vec<f32>, FetchError> {
    let url = format!("https://www.ote-cr.cz/en/short-term-markets/electricity/day-ahead-market/@@chart-data?report_date={}", date);
    info!("Fetching data for date {}", date);

    let start = std::time::Instant::now();

    // Create a client
    let client = Client::new();

    // Make the GET request
    let response = client.get(&url).send().await.map_err(|error| {
        error!(
            "Request failed {} in {:?} error {}",
            date,
            start.elapsed(),
            error
        );

        error
    })?;

    info!(
        "Request for {} in {:?} status {}",
        date,
        start.elapsed(),
        response.status()
    );

    // Check if the request was successful
    if response.status().is_success() {
        // Parse the response JSON
        let response_json: Response = response.json().await?;

        // Find the data line with the title "Price (EUR/MWh)"
        if let Some(price_data) = response_json
            .data
            .data_line
            .iter()
            .find(|line| line.title == "15min price (EUR/MWh)")
        {
            // Check if the number of points is exactly 24
            if price_data.point.len() != 96 {
                error!("Price data does not contain exactly 96 points.");
                return Err(FetchError::InvalidDataSize);
            }

            Ok(price_data.point.iter().map(|point| point.y).collect())
        } else {
            error!("Price data not found in the response.");
            Err(FetchError::PriceDataNotFound)
        }
    } else {
        error!("Failed to fetch data. Status: {}", response.status());
        Err(FetchError::UnexpectedStatus(response.status()))
    }
}

File: dataloader/dataloader.go
package dataloader

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net/http"
	"time"
)

type point struct {
	Y float32 `json:"y"`
}

type dataLine struct {
	Title string  `json:"title"`
	Point []point `json:"point"`
}

type data struct {
	DataLine []dataLine `json:"dataLine"`
}

type response struct {
	Data data `json:"data"`
}

var (
	ErrPriceDataNotFound = errors.New("Price data not found")
	ErrInvalidDataSize   = errors.New("Price data does not contain exactly 24 points")
)

type UnexpectedStatusError struct {
	Status int
}

func (e *UnexpectedStatusError) Error() string {
	return fmt.Sprintf("Unexpected response status: %d", e.Status)
}

// FetchData fetches day-ahead electricity prices for the given date.
func FetchData(date time.Time) ([]float32, error) {
	dateStr := date.Format("2006-01-02")
	url := fmt.Sprintf(
		"https://www.ote-cr.cz/en/short-term-markets/electricity/day-ahead-market/@@chart-data?report_date=%s",
		dateStr,
	)
	log.Printf("Fetching data for date %s", dateStr)

	start := time.Now()

	client := &http.Client{}
	req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("Network error: %w", err)
	}

	resp, err := client.Do(req)
	if err != nil {
		log.Printf("Request failed %s in %v error %v", dateStr, time.Since(start), err)
		return nil, fmt.Errorf("Network error: %w", err)
	}
	defer resp.Body.Close()

	log.Printf("Request for %s in %v status %d", dateStr, time.Since(start), resp.StatusCode)

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		log.Printf("Failed to fetch data. Status: %d", resp.StatusCode)
		return nil, &UnexpectedStatusError{Status: resp.StatusCode}
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("Network error: %w", err)
	}

	var respJSON response
	if err := json.Unmarshal(body, &respJSON); err != nil {
		return nil, fmt.Errorf("JSON parsing error: %w", err)
	}

	for _, line := range respJSON.Data.DataLine {
		if line.Title == "15min price (EUR/MWh)" {
			if len(line.Point) != 96 {
				log.Printf("Price data does not contain exactly 96 points.")
				return nil, ErrInvalidDataSize
			}
			out := make([]float32, len(line.Point))
			for i, p := range line.Point {
				out[i] = p.Y
			}
			return out, nil
		}
	}

	log.Printf("Price data not found in the response.")
	return nil, ErrPriceDataNotFound
}
